rsyslog/undefined_field/undefined_field.go

```diff
@@ -205,6 +205,7 @@ func processUndefinedAndMaxNumFields(input map[string]interface{}) (string, map[
 			}
 			undefMap[field] = val
 			count--
+			if count < 0 { break }
```

Comment:
There may be more undefined fields.  For example, if the `cfg.UndefinedMaxNumFields` is 2, and there are 4 undefined fields in the record, we want `undefMap` to contain all 4 of them, so that when we convert `undefMap` to `undefString`, it will contain all of the undefined fields.

Reply:
> There may be more undefined fields. For example, if the `cfg.UndefinedMaxNumFields` is 2, and there are 4 undefined fields in the record, we want `undefMap` to contain all 4 of them, so that when we convert `undefMap` to `undefString`, it will contain all of the undefined fields.

@richm, I might still misunderstand the definition of processUndefinedAndMaxNumFields, but I still think the function should return 2 undefString and remove 4 fields from input (in your example).  I've updated the patch according to it and it passes the unit test.  Can I have your thought?

Reply:
it should work like this: https://github.com/ViaQ/fluent-plugin-viaq_data_model
```
If undefined_max_num_fields is a number greater than -1, and if the number of undefined fields is greater than this number, all of the undefined fields will be converted to their JSON string representation and stored in the undefined_name named field.
...
undefined_max_num_fields - integer - default -1

    If the number of undefined fields exceeds the value of undefined_max_num_fields, then convert the hash of undefined fields to its JSON string representation, and store the values in the undefined_name field - see below
    Use a value of -1 if you want to have an unlimited number of undefined fields (not recommended)
    Using undefined_max_num_fields implies that you want to use undefined_name as the name of the field to store the value, even if use_undefined is not set - if you want to use a different field name than "undefined" then set undefined_name

```
We have to find all of the undefined fields so we can convert all of them to string.